docs/proposals/20230123-azure-service-operator.md

```diff
@@ -0,0 +1,188 @@
+
+### Implementation Details/Notes/Constraints
+
+- ASO needs to be installed on the management cluster
+  - Multi- or single-tenant?
+  - Install in capz-controller-manager `func main()`
+- GET requests are not as straightforward in the ASO model because BYO resources
+  may not initially exist in ASO
+- Current service pattern in CAPZ works for Azure services that are not
+  represented in ASO (resource health)
+- Each release of ASO can expose several Azure API versions of each resource,
+  including GA and preview API versions.
+- Must ensure CAPZ only operates against the ASO resources it manages, not ones managed by the user.
+
+### Security Model
+
+ASO requires credentials to interact with Azure which are documented here: https://azure.github.io/azure-service-operator/introduction/authentication/. Specifically, ASO needs the values of `$AZURE_CLIENT_ID`, `$AZURE_CLIENT_SECRET`, `$AZURE_SUBSCRIPTION_ID`, and `$AZURE_TENANT_ID` for Service Principal authentication, which are already set in the capz-controller-manager container's environment variables. Passing them along as chart values when installing ASO with Helm doesn't introduce any significant additional security risks.
+
+Consumers of ASO's APIs (such as CAPZ) also need Kubernetes RBAC permissions to interact with ASO resources. For example, this ClusterRole snippet is required by CAPZ to manage resource groups with ASO:
+```yaml
+- apiGroups:
+  - resources.azure.com
+  resources:
+  - resourcegroups
+  - resourcegroups/status
+  verbs:
+  - get
+  - list
+  - watch
+  - create
+  - update
+  - delete
+```
+The above permissions may overlap with ASO resources already installed by a user not intended to be managed by CAPZ.
+
+### Risks and Mitigations
+
+- Introduction of functional regressions
+- Introduction of performance regressions, where operations against workload
+  clusters may take longer to complete
+- Introduction of scalability regressions, where management clusters may exhaust
+  Azure or Kubernetes API limits with fewer or smaller workload clusters being managed.
+  - Management cluster will have to manage many more Kubernetes resources per
+    workload cluster
+- Because ASO has not yet been proven as a mission-critical interface to Azure
+  to the same extent as Azure's REST API and Go SDK, there is an apparent increased risk
+  that CAPZ's lifespan may extend beyond that of its Azure interface when ASO is
+  used instead of the API or SDK directly
+- Conflicting user installations of ASO or ASO resources
+- Future breaking changes in ASO
+- Lower-fidelity telemetry compared to what CAPZ tracks currently
```

Comment:
This is something we'd love to work with you guys on I think. We have some basic telemetry exposed already: https://azure.github.io/azure-service-operator/introduction/metrics/ - if you gave us a list of what exactly you wanted (or were losing in this migration) we could work to expose that data.

Or is the issue here more than you had integrations into the Azure SDK to track aggregate metrics such as "time it takes to fully provision a cluster" that you'd be losing?

Reply:
For the little bit I've used CAPZ's tracing, I've found it helpful to have a breakdown of how long each step in a single CAPZ reconciliation takes. Since that includes Azure API calls currently, I think my main concern was losing that kind of association between a CAPZ reconciliation and Azure API calls. I updated this section to mention that I don't think that would really matter though since Azure API calls would be happening in ASO completely out-of-band with CAPZ reconciliations. Or at least recreating that mapping seems like it would be unnecessarily difficult.

Reply:
There have been discussions about tracking resource lifecycle and some related KEP work: https://groups.google.com/g/kubebuilder/c/tNI6ZpQ2loM/m/8rSX6HKVDgAJ. Correlation is going to be difficult. However, we might be able to trace with observed generation and namespace/name to get something close enough.